File: JavaScript/Basics/Function/program1.js
// Function (Definitions)

// Javascript functions are defined with the 'function' keyword

// Function declarations
function myfunction(num1, num2) {
    return num1*num2;
}

console.log(myfunction(10, 20)); // 200 (call / invoke a function)

// Function Expressions
const myfunction2 = function (a, b) {return a * b}; // The function above is actually an anonymous function(a function without a name)

// Functions stored in variables do not need function names. They are always invoked (called) using the variable name

let callfunction = myfunction2(1,3); // 3 (after a function expression has been stored in a variable, the variable can be used as a function)

// Functions can also be defined with a built-in Javascript function constructor called 'Function()'

const myfunction3 = new Function('a', 'b', 'return a*b');
let x = myfunction3(4,3); // 12

// ***Most of the time, we avoid using 'new' keyword in Javascript

// Self-Invoking functions -> is invoked(started) automatically, without being called

(function(){
    console.log("Hello world!"); 
    
})(); // Hello world

// Functions can be used as values
function func(a, b){
    return a*b;
}

let y = func(4,3); // 12
console.log(y);

y = func(4,3) * 2; // 24
console.log(y);

// Arrow functions -> allows a short syntax for writing function expressions

const arrowFunc = (a, b)=>{
    return a*b; // (a, b) => a*b (for one line avoid return keyword)
} 

console.log(arrowFunc(10,30)); // 300

File: C/Basics/Function/program1.c
#include <stdio.h>

// int add(int num1, int num2) 
double add(double num1, double num2)
{
    double sum = num1 + num2;
    return sum;
}
 
int main()
{
    // defining a function 
    // return_type function_name(parameters){
    //  function body
    //  return value
    // }
    
    // function_name(arguments) // invoke or call a function

    double a = 2.5, b = 2.5, c;
    c = a + b;
    printf("%lf\n", c); // 5.000000

    c = add(2.5, 2.5);
    printf("%lf\n", c); // 4.000000

    return 0;
}   